dbms/src/Interpreters/JoinV2/HashJoinProbe.cpp

```diff
@@ -0,0 +1,666 @@
+        ASTTableJoin::Kind kind,
+        const JoinNonEqualConditions & non_equal_conditions,
+        const HashJoinSettings & settings,
+        const HashJoinPointerTable & pointer_table,
+        const HashJoinRowLayout & row_layout,
+        MutableColumns & added_columns,
+        size_t added_rows)
+        : context(context)
+        , wd(wd)
+        , method(method)
+        , kind(kind)
+        , non_equal_conditions(non_equal_conditions)
+        , settings(settings)
+        , pointer_table(pointer_table)
+        , row_layout(row_layout)
+        , added_columns(added_columns)
+        , added_rows(added_rows)
+    {
+        wd.insert_batch.clear();
+        wd.insert_batch.reserve(settings.probe_insert_batch_size);
+
+        wd.selective_offsets.clear();
+        wd.selective_offsets.reserve(settings.max_block_size);
+
+        if (pointer_table.enableProbePrefetch() && !context.prefetch_states)
+        {
+            context.prefetch_states = decltype(context.prefetch_states)(
+                static_cast<void *>(new ProbePrefetchState<KeyGetter>[settings.probe_prefetch_step]),
+                [](void * ptr) { delete static_cast<ProbePrefetchState<KeyGetter> *>(ptr); });
+        }
+    }
+
+    void joinProbeBlockImpl();
+
+    void NO_INLINE joinProbeBlockInner();
+    void NO_INLINE joinProbeBlockInnerPrefetch();
+
+    void NO_INLINE joinProbeBlockLeftOuter();
+    void NO_INLINE joinProbeBlockLeftOuterPrefetch();
+
+    void NO_INLINE joinProbeBlockSemi();
+    void NO_INLINE joinProbeBlockSemiPrefetch();
+
+    void NO_INLINE joinProbeBlockAnti();
+    void NO_INLINE joinProbeBlockAntiPrefetch();
+
+    template <bool has_other_condition>
+    void NO_INLINE joinProbeBlockRightOuter();
+    template <bool has_other_condition>
+    void NO_INLINE joinProbeBlockRightOuterPrefetch();
+
+    template <bool has_other_condition>
+    void NO_INLINE joinProbeBlockRightSemi();
+    template <bool has_other_condition>
+    void NO_INLINE joinProbeBlockRightSemiPrefetch();
+
+    template <bool has_other_condition>
+    void NO_INLINE joinProbeBlockRightAnti();
+    template <bool has_other_condition>
+    void NO_INLINE joinProbeBlockRightAntiPrefetch();
+
+private:
+    bool ALWAYS_INLINE joinKeyIsEqual(
+        KeyGetterType & key_getter,
+        const KeyType & key1,
+        const KeyType & key2,
+        HashValueType hash1,
+        RowPtr row_ptr) const
+    {
+        if constexpr (KeyGetterType::joinKeyCompareHashFirst())
+        {
+            auto hash2 = unalignedLoad<HashValueType>(row_ptr + sizeof(RowPtr));
+            if (hash1 != hash2)
+                return false;
+        }
+        return key_getter.joinKeyIsEqual(key1, key2);
+    }
+
+    void ALWAYS_INLINE insertRowToBatch(KeyGetterType & key_getter, RowPtr row_ptr, const KeyType & key) const
+    {
+        wd.insert_batch.push_back(row_ptr + key_getter.getRequiredKeyOffset(key));
+        FlushBatchIfNecessary<false>();
```

Comment:
Invalid code style, should be `flushBatchIfNecessary()` ?

Reply:
Indeed. Fixed.